File: 0200-number-of-islands/0200-number-of-islands.java
class Solution {
    public int numIslands(char[][] grid) {
        int m = grid.length;
        int n = grid[0].length;
        Queue<Pair<Integer, Integer>> queue = new LinkedList<>();
        int total = 0;
        int[][] nei = {{1, 0}, {0, 1}, {-1, 0}, {0, -1}};

        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (grid[i][j] == '0') continue;
                queue.offer(new Pair<Integer, Integer>(i, j));
                grid[i][j] = '0';
                while (!queue.isEmpty()) {
                    var top = queue.poll();
                    for (var idx : nei) {
                        var x = top.getKey() + idx[0];
                        var y = top.getValue() + idx[1];
                        if (x < 0 || x >= m || y < 0 || y >= n || grid[x][y] != '1') continue;
                        queue.offer(new Pair<Integer, Integer>(x, y));
                        grid[x][y] = '0';
                    }
                }
                total++;
            }
        }

        return total;
    }
}

File: 0200-number-of-islands/0200-number-of-islands.cpp
class Solution {
constexpr static int directions[4][2] = {{0, 1}, {1, 0}, {-1, 0}, {0, -1}};

public:
    int numIslands(vector<vector<char>>& grid) {
        int m = grid.size(), n = grid[0].size();
        queue<pair<int, int>> q;
        int islands{};

        for (auto i{0uz}; i < m; ++i) {
            for (auto j{0uz}; j < n; ++j) {
                if (grid[i][j] != '1') continue;
                q.push({i, j});
                while (!q.empty()) {
                    auto [x, y] = q.front();
                    q.pop();
                    grid[x][y] = '2';
                    for (auto d : directions) {
                        auto x1 = x + d[0];
                        auto y1 = y + d[1];
                        if (x1 < 0 || x1 >= m || y1 < 0 || y1 >= n || grid[x1][y1] != '1') continue;
                        grid[x1][y1] = '2';
                        q.push({x1, y1}); 
                    }
                }

                ++islands;
            }
        }

        return islands;        
    }
};